use visitor for #[structural_match] check

pnkfelix opened the pull request:
This changes the code so that we recur down the structure of a type of a const (rather than just inspecting at a shallow one or two levels) when we are looking to see if it has an ADT that did not derive `PartialEq` and `Eq`.

Fix #61188 

Fix #62307 

Cc #62336

rust-highfive commented:
r? @varkor

(rust_highfive has picked a reviewer for you, use r? to override)

pnkfelix commented:
The way this is written might be bad for performance, since it is doing a visit of the type structure on every call to `const_to_pat`, and that routine calls itself recursively.

Thinking further on this, we *might* be able to avoid that cost by factoring the recursive traversal into a subroutine, and then doing the type traversal once, at the start of `const_to_pat`, before starting the recursive traversal. (But I would want to be careful about not overlooking cases where the problematic ADT occurs more deeply in the input.)

pnkfelix commented:
(I've left the WIP marker because I want to add tests to double-check that the refactoring in commit 9db458f300a1f0c844071a41a547bd6294c99610 didn't break things.)

Centril commented on src/librustc_mir/hair/pattern/mod.rs:
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
     }
 }
 
+fn search_for_adt_without_structural_match<'tcx>(tcx: TyCtxt<'tcx>,
```
Indentation is weird here, should be:
```rust
fn search_for_adt_without_structural_match<'tcx>(
    tcx: TyCtxt<'tcx>,
    ty: Ty<'tcx>,
) -> Option<&'tcx AdtDef> {
    ...
```

Centril commented on src/librustc_mir/hair/pattern/mod.rs:
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
+
+                // For ADT, we do *not* visit contents.
+                //
+                // The presence of the #[structural_match] attribute
```
```suggestion
                // The presence of the `#[structural_match]` attribute
```

Centril commented on src/librustc_mir/hair/pattern/mod.rs:
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
+                if !self.tcx.has_attr(adt_def.did, sym::structural_match) {
+                    self.found = Some(&adt_def);
+                    debug!("SearchForAdtWithoutStructuralMatch found adt_def: {:?}", adt_def);
+                    return true // Halt visiting
```
```suggestion
                    return true // Halt visiting!
```

Centril commented on src/librustc_mir/hair/pattern/mod.rs:
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
+                // (Also, visiting contents would visit the
+                // substitutions of any type parameters on the ADT,
+                // which we do not want. Visitng them mishandles
+                // PhantomData<T> according to rust-lang/rust#55028
```
```suggestion
                // `PhantomData<T>` according to rust-lang/rust#55028
```

Centril commented on src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs:
```diff
@@ -0,0 +1,19 @@
+// Issue 61118 pointed out a case where we hit an ICE during code gen:
+// the compiler assumed that PartialEq was always implemented on any
```
```suggestion
// the compiler assumed that `PartialEq` was always implemented on any
```

Centril commented on src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs:
```diff
@@ -0,0 +1,19 @@
+// Issue 61118 pointed out a case where we hit an ICE during code gen:
+// the compiler assumed that PartialEq was always implemented on any
+// const used in a pattern context, but the code responsible for
```
```suggestion
// `const` used in a pattern context, but the code responsible for
```

pnkfelix replied:
mmm, I usually don't put something in `code format` when I'm not using the syntactic form and am instead using the word to describe a mechanism or broader pattern. Sort of like how one would not put "trait" in `code format` unless one were talking about a bit of code that actually used the `trait` keyword.

Centril replied:
Fair :)

pnkfelix replied:
Would you prefer I use another word here? I did consider saying "constant" but then I worried that would then stray into the territory of describing a syntactic literal, and I need the reader to keep in mind that we are talking about uses of `const` items.

Maybe I should say it that way, "... that `PartialEq` was always implemented on any use of a `const` item in a pattern context, but ..." ?

Centril replied:
I wouldn't overthink it -- it was mostly a nit on my part. ;)
That said, your last suggestion makes sense.

Centril commented on src/test/ui/rfc1445/issue-62307-match-ref-ref-forbidden-without-eq.rs:
```diff
@@ -0,0 +1,39 @@
+// RFC 1445 introduced `#[structural_match]`; this attribute must
+// appear on the struct/enum definition for any const used in a
```
```suggestion
// appear on the `struct`/`enum` definition for any `const` used in a
```

Centril commented on src/test/ui/rfc1445/issue-62307-match-ref-ref-forbidden-without-eq.rs:
```diff
@@ -0,0 +1,39 @@
+#[derive(Debug)]
+struct B(i32);
+
+// Overriding PartialEq to use this strange notion of "equality" exposes
```
```suggestion
// Overriding `PartialEq` to use this strange notion of "equality" exposes
```

Centril commented on src/test/ui/rfc1445/match-nonempty-array-forbidden-without-eq.rs:
```diff
@@ -0,0 +1,19 @@
+// `#[structural_match]` was too shallow.
+//
+// Here we check similar behavior for non-empty arrays of types that
+// do not derive Eq.
```
```suggestion
// do not derive `Eq`.
```

matthewjasper commented:
Does this handle matching on `const A: &[Option<NoDerive>] = ...`?

pnkfelix commented:
@matthewjasper no, it mishandles the case of `const A: &[Option<NoDerive>]` (as in, it fails to reject a pattern using that const).

The code also mishandles some other cases that I just discovered.
 * I wanted to double check the text of rust-lang/rfcs#1445 to make sure I understand the intended scope of that change
 * its possible the RFC *intended* for us to use a shallow check of the attribute; i.e., that we would accept the case of a struct wrapping another struct, where the wrapper derives `PartialEq` and `Eq` but the wrapped type does not derive them and instead uses its own implementation.
 * But I don't think that could have been the true intention, since allowing such side-stepping of the attribute via a wrapper struct seems like it would undermine the objectives of the RFC itself.
 * (I *suspect* the reality is the RFC was written assuming a sort of recursive semantics, where converting the wrapper to a pattern would involve converting the wrapped struct to a pattern, which in turn would cause the wrapped struct to be checked. But this is not what currently happens, due to the way we sometimes dispatch to a `PartialEq::eq` call rather than expanding.)

I'll post a follow-up comment outlining the behavior on a set of cases after I finishing going through the analysis.

pnkfelix commented:
Okay, that last change (to move the visit up above the recursive loop) broke cases like these:

```rust
struct NoDerive(i32);

impl PartialEq for NoDerive {
    fn eq(&self, _: &Self) -> bool { false }
}

impl Eq for NoDerive { }

#[derive(PartialEq, Eq)]
struct WrapInline(NoDerive);

const WRAP_DIRECT_INLINE: WrapInline = WrapInline(NoDerive(0));

#[derive(PartialEq, Eq)]
struct WrapParam<T>(T);

const WRAP_DIRECT_PARAM: WrapParam<NoDerive> = WrapParam(NoDerive(0));
```

Meanwhile, regardless of where the visit is done, I also fail to correctly handle:

```rust
const WRAP_INDIRECT_INLINE: & &WrapInline = & &WrapInline(NoDerive(0));

const WRAP_INDIRECT_PARAM: & &WrapParam<NoDerive> = & &WrapParam(NoDerive(0));
```

pnkfelix commented:
Okay so **now** I've removed commit 9db458f300a1f0c844071a41a547bd6294c99610 to avoid regressing cases that we *were* handling correctly before, and I've added 1e55df7a3ab2fca0ec35b884cf556af4b76d3624 because I've decided that we do need to visit the substs of an ADT (and just treat `PhantomData<T>` as a special case).

With that change in place, we handle @matthewjasper 's example correctly, as well as most of the other ones I identified.

The one case that this version continues to mishandle is this ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9fb46f1f29d393cc7bdf4568e07b7d65)):

```rust
struct NoDerive;

impl PartialEq for NoDerive {
    fn eq(&self, _: &Self) -> bool { false }
}

impl Eq for NoDerive { }

#[derive(PartialEq, Eq)]
struct WrapInline(NoDerive);

const WRAP_INLINE: & &WrapInline = & &WrapInline(NoDerive);

fn main() {
    match WRAP_INLINE {
        WRAP_INLINE => { println!("WRAP_INLINE matched WRAP_INLINE"); }
        _ => { println!("WRAP_INLINE did not match WRAP_INLINE"); }
    }
}
```

I think handling this case correctly (and also efficiently) will require deeper changes, though. E.g. a new bit in `rustc::ty::AdtFlags`.

 * To be honest we might be better off trying to address finding our long-term solution to the structural-match problem rather than trying to plug all the leaks in `#[structural_match]`...
 * But under the assumption that gathering consensus on a language problem like that will always take more time than you expect, I'll see about changing `rustc::ty::AdtFlags` in the short term.

pnkfelix commented:
Okay, so *now* I've added a more complete traversal that addresses the problem posted above.

But doing that then led me to ask some questions, like: 
 * "Now that I am traversing the fields, should I stop traversing the substs of an ADT?"
   * (probably yes)
 * "How should unsafe pointers be handled?" 
   * (probably should not recur on the `T` in `*const T`)

The fact that I'm asking these questions leads me to think three things:
  * I need to add more tests
  * The implementation here needs more work
  * This probably needs to be initially deployed with a warning cycle. (I don't think that will be too hard to do, even.)

Centril commented:
r? @nikomatsakis  since they wrote the original RFC.

pnkfelix commented:
Turning this from an error to a warning exposes *more* issues; e.g. the point where we are doing this check in the compiler's control flow, namely lowering from HIR to HAIR, is a spot where for some reason we may attempt to lower the same input multiple times, and thus the warning gets issued redundantly.

(The very fact that we are re-running lower on the same input seems like a worrisome thing, just in terms of compiler efficiency...)

So, do I focus on making this a `LintPass` rather than doing it in lowering? Or do I try to figure out why lowering is being run multiple times on the same input?

pnkfelix commented:
Okay I'm finally narrowing in on a solution that I'm happy with, I hope.

I am going to try to deploy this as a future-compat lint; I've filed #62411 as a tracking issue for that.

pnkfelix commented:
(last commit 02af3ca5c42f35eeceb0761180193a6abb97c0ad was result of my realization that while I had added [special case code](https://github.com/rust-lang/rust/pull/62339/commits/b56080162bc45e1cfd561b0a21e50b827c9df679#diff-e826542d6f24137d5b0b5ae333230529R1185) to preserve existing behavor for empty-arrays (#62336), I hadn't put in a corresponding unit test. That's fixed now.)

nikomatsakis commented on src/librustc_mir/hair/pattern/mod.rs:
```diff
@@ -1058,6 +1145,96 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
 
+fn search_for_adt_without_structural_match<'tcx>(tcx: TyCtxt<'tcx>,
+                                                 ty: Ty<'tcx>)
+                                                 -> Option<&'tcx AdtDef>
```
I think this function could use a comment giving more context and perhaps pointing to some of the relevant issues. It's name is good but would be pretty opaque if you didn't understand what was going on.

pnkfelix replied:
Good idea, added such a comment in b0b64dd

nikomatsakis commented:
My 2 cents:

- r=me on the PR itself.
- it seems like a good idea to do a crater run (with this lint level set to deny) to get an idea of the impact.
- I do think we ought to resolve this question around constants. This feels like something where, if we had a functional working group around const generics, they could ultimately resolve it.

pnkfelix commented:
hmm, looking at #36891, I now wonder: should we use a different name than `non_structural_match` for the future-compat lint here? I had considered e.g. `indirect_non_structural_match`...

Having said that, I nonetheless plan to let this PR land as is. We can bikeshed the names while the PR is on nightly (right?)

pnkfelix commented:
(actually on second thought I probably will do some rename of the lint before landing it. the current name is a bit too broad for my taste...)

pnkfelix commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit b0b64ddcd171be6e460b2d7c1ada4f98a3cdb026 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis b0b64ddcd171be6e460b2d7c1ada4f98a3cdb026 -->
<!-- homu: {"type":"Approved","sha":"b0b64ddcd171be6e460b2d7c1ada4f98a3cdb026","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit b0b64ddcd171be6e460b2d7c1ada4f98a3cdb026 with merge c6a9e766f90a5271c2356fbc7941e38559200ab5...
<!-- homu: {"type":"BuildStarted","head_sha":"b0b64ddcd171be6e460b2d7c1ada4f98a3cdb026","merge_sha":"c6a9e766f90a5271c2356fbc7941e38559200ab5"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=3051), [checks-travis](https://travis-ci.com/rust-lang/rust/builds/118600612), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/25876394)
Approved by: nikomatsakis
Pushing c6a9e766f90a5271c2356fbc7941e38559200ab5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang/rust/builds/25876394","checks-travis":"https://travis-ci.com/rust-lang/rust/builds/118600612","checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=3051"},"merge_sha":"c6a9e766f90a5271c2356fbc7941e38559200ab5"} -->
